Modules/FindCUDA.cmake

```diff
@@ -586,6 +587,49 @@ if(NOT "${CUDA_SDK_ROOT_DIR}" STREQUAL "${CUDA_SDK_ROOT_DIR_INTERNAL}")
   #   unset(MY_SPECIAL_CUDA_SDK_LIBRARY CACHE)
 endif()
 
+#
+#  End of unset()
+#
+
+#
+#  Start looking for things
+#  Before we do, find out who we are
+#
+
+if(CMAKE_CROSSCOMPILING)
+  SET (CUDA_TOOLKIT_ROOT $ENV{CUDA_TOOLKIT_ROOT})
+  if(CMAKE_SYSTEM_PROCESSOR STREQUAL "armv7-a")
+    # Support for NVPACK
+    set (CUDA_TOOLKIT_TARGET_NAME "armv7-linux-androideabi")
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "arm")
+    # Support for arm cross compilation
+    set(CUDA_TOOLKIT_TARGET_NAME "armv7-linux-gnueabihf")
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64")
+    # Support for aarch64 cross compilation
+    if (ANDROID_ARCH_NAME STREQUAL "arm64")
+      set(CUDA_TOOLKIT_TARGET_NAME "aarch64-linux-androideabi")
+    else()
+      set(CUDA_TOOLKIT_TARGET_NAME "aarch64-linux")
+    endif (ANDROID_ARCH_NAME STREQUAL "arm64")
+  endif()
+
+  if (EXISTS "${CUDA_TOOLKIT_ROOT}/targets/${CUDA_TOOLKIT_TARGET_NAME}")
+    set(CUDA_TOOLKIT_TARGET_DIR "${CUDA_TOOLKIT_ROOT}/targets/${CUDA_TOOLKIT_TARGET_NAME}" CACHE PATH "CUDA Toolkit target location.")
+    SET (CUDA_TOOLKIT_ROOT_DIR ${CUDA_TOOLKIT_ROOT})
+    mark_as_advanced(CUDA_TOOLKIT_TARGET_DIR)
+  endif()
+
+  # Target CPU architecture: this nvcc option is now obsolete
+  # add known CUDA targetr root path to the set of directories we search for programs, libraries and headers
+  set( CMAKE_FIND_ROOT_PATH "${CUDA_TOOLKIT_TARGET_DIR};${CMAKE_FIND_ROOT_PATH}")
```

Comment:
In fact, I read the documentation: https://cmake.org/cmake/help/v3.0/manual/cmake-toolchains.7.html?highlight=cmake_find_root_path

In your case, yes, having CUDA_TOOLKIT_ROOT_DIR in CMAKE_FIND_ROOT_PATH and adding CUDA_TOOLKIT_TARGET_DIR works, since CUDA_TOOLKIT_TARGET_DIR is under CUDA_TOOLKIT_ROOT_DIR Vice versa - does not work unless one changes the setting defined by the toolchain cmake files: 

# only search for libraries and includes in the ndk toolchain                                                                                                  
set( CMAKE_FIND_ROOT_PATH_MODE_PROGRAM ONLY )
set( CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY )
set( CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY )

In NVPACK case, CUDA root is outside of Android sysroot set by toolchain file.
If I change 'ONLY' to both then yes I will be able to find the packages outside the target sysroot, but then I will risk to find host (and other targets)  CUDA library instead of the one I was looking for.
So the right thing to do here is exactly what I did: if target's CUDA library is being installed outside of the known target system "root", add the target CUDA directory to the set of roots the search is being restricted to.